File: sdks/js/examples/chat_app.ts
/**
 * Chat Application Example
 * 
 * Demonstrates a multi-actor chat room application with room and session actors.
 */
import { Actor, Message, MessageType, State } from '@aether/sdk';

// ============================================
// Types
// ============================================

interface User {
    user_id: string;
    name: string;
    joined_at: string;
}

interface ChatMessage {
    id: string;
    user_id: string;
    user_name: string;
    content: string;
    timestamp: string;
    room_id: string;
}

interface RoomState {
    room_id: string;
    room_name: string;
    users: User[];
    messages: ChatMessage[];
}

// ============================================
// Room Actor
// ============================================

class RoomActor extends Actor {
    private roomId: string;
    private roomName: string;
    private users: Map<string, User> = new Map();
    private messages: ChatMessage[] = [];
    private stateKey: string;
    private state: State;

    constructor(roomId: string, roomName: string) {
        super(`room-${roomId}`);
        this.roomId = roomId;
        this.roomName = roomName;
        this.stateKey = `room_${roomId}_state`;
        this.state = new State();
        this.require('STATE_READ', 'STATE_WRITE', 'ACTOR_MESSAGING', 'LOG', 'TIME');
    }

    async onStart(): Promise<void> {
        console.log(`[${this.name}] Room '${this.roomName}' starting...`);

        // Load persisted state
        try {
            const data = await this.state.read(this.stateKey);
            if (data) {
                const state: RoomState = JSON.parse(data);
                
                for (const u of state.users || []) {
                    this.users.set(u.user_id, u);
                }
                
                for (const m of state.messages || []) {
                    this.messages.push(m);
                }
                
                console.log(`[${this.name}] Restored ${this.users.size} users, ${this.messages.length} messages`);
            } else {
                await this.saveState();
                console.log(`[${this.name}] Initialized new room`);
            }
        } catch (error) {
            console.error(`[${this.name}] Failed to load state:`, error);
            await this.saveState();
        }
    }

    async onStop(): Promise<void> {
        console.log(`[${this.name}] Room stopping, saving state...`);
        await this.saveState();
    }

    async handleMessage(sender: string, message: Message): Promise<Message | null> {
        if (message.type !== MessageType.REQUEST && message.type !== MessageType.RPC_REQUEST) {
            return null;
        }

        const payload = message.payload as Record<string, any> | null;
        if (!payload || typeof payload !== 'object') {
            return Message.response({ error: 'invalid payload' });
        }

        const action = payload.action || '';

        switch (action) {
            case 'join':
                return this.handleJoin(payload);
            case 'leave':
                return this.handleLeave(payload);
            case 'send':
                return this.handleSend(payload);
            case 'history':
                return this.handleHistory(payload);
            case 'users':
                return this.handleUsers();
            case 'info':
                return this.handleInfo();
            default:
                return Message.response({ error: `unknown action: ${action}` });
        }
    }

    private async handleJoin(payload: Record<string, any>): Promise<Message> {
        const userId = payload.user_id || '';
        const userName = payload.user_name || '';

        if (!userId || !userName) {
            return Message.response({ error: 'user_id and user_name required' });
        }

        const user: User = {
            user_id: userId,
            name: userName,
            joined_at: new Date().toISOString()
        };

        this.users.set(userId, user);
        await this.saveState();

        console.log(`[${this.name}] User '${userName}' joined (total: ${this.users.size})`);

        return Message.response({
            action: 'joined',
            room_id: this.roomId,
            room_name: this.roomName,
            user_id: userId,
            user_count: this.users.size
        });
    }

    private async handleLeave(payload: Record<string, any>): Promise<Message> {
        const userId = payload.user_id || '';

        const user = this.users.get(userId);
        if (user) {
            this.users.delete(userId);
            await this.saveState();
            console.log(`[${this.name}] User '${user.name}' left (remaining: ${this.users.size})`);
        }

        return Message.response({
            action: 'left',
            room_id: this.roomId,
            user_id: userId,
            user_count: this.users.size
        });
    }

    private async handleSend(payload: Record<string, any>): Promise<Message> {
        const userId = payload.user_id || '';
        const content = payload.content || '';

        if (!userId || !content) {
            return Message.response({ error: 'user_id and content required' });
        }

        const user = this.users.get(userId);
        if (!user) {
            return Message.response({ error: 'user not in room, join first' });
        }

        const msg: ChatMessage = {
            id: `msg-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
            user_id: userId,
            user_name: user.name,
            content: content,
            timestamp: new Date().toISOString(),
            room_id: this.roomId
        };

        this.messages.push(msg);
        await this.saveState();

        console.log(`[${this.name}] [${msg.id.substring(0, 12)}] ${user.name}: ${content.substring(0, 30)}...`);

        return Message.response({
            action: 'sent',
            message_id: msg.id,
            message_count: this.messages.length
        });
    }

    private handleHistory(payload: Record<string, any>): Message {
        const limit = payload.limit || 50;
        const start = Math.max(0, this.messages.length - limit);
        const messages = this.messages.slice(start);

        return Message.response({
            action: 'history',
            room_id: this.roomId,
            messages: messages,
            count: messages.length
        });
    }

    private handleUsers(): Message {
        const users = Array.from(this.users.values());
        return Message.response({
            action: 'users',
            room_id: this.roomId,
            users: users,
            count: users.length
        });
    }

    private handleInfo(): Message {
        return Message.response({
            action: 'info',
            room_id: this.roomId,
            room_name: this.roomName,
            user_count: this.users.size,
            message_count: this.messages.length
        });
    }

    private async saveState(): Promise<void> {
        const state: RoomState = {
            room_id: this.roomId,
            room_name: this.roomName,
            users: Array.from(this.users.values()),
            messages: this.messages
        };
        await this.state.write(this.stateKey, JSON.stringify(state));
    }
}

// ============================================
// Session Actor
// ============================================

class SessionActor extends Actor {
    private userId: string;
    private userName: string;
    private rooms: Set<string> = new Set();

    constructor(userId: string, userName: string) {
        super(`session-${userId}`);
        this.userId = userId;
        this.userName = userName;
        this.require('ACTOR_MESSAGING', 'LOG');
    }

    async onStart(): Promise<void> {
        console.log(`[${this.name}] Session started for user '${this.userName}'`);
    }

    async onStop(): Promise<void> {
        console.log(`[${this.name}] Session ended for user '${this.userName}'`);
    }

    async handleMessage(sender: string, message: Message): Promise<Message | null> {
        if (message.type !== MessageType.REQUEST && message.type !== MessageType.RPC_REQUEST) {
            return null;
        }

        const payload = message.payload as Record<string, any> | null;
        if (!payload || typeof payload !== 'object') {
            return Message.response({ error: 'invalid payload' });
        }

        const action = payload.action || '';

        switch (action) {
            case 'status':
                return Message.response({
                    action: 'status',
                    user_id: this.userId,
                    user_name: this.userName,
                    rooms: Array.from(this.rooms)
                });
            case 'join_room':
                return this.handleJoinRoom(payload);
            case 'leave_room':
                return this.handleLeaveRoom(payload);
            default:
                return Message.response({ error: `unknown action: ${action}` });
        }
    }

    private handleJoinRoom(payload: Record<string, any>): Message {
        const roomId = payload.room_id || '';
        this.rooms.add(roomId);
        console.log(`[${this.name}] Joined room '${roomId}' (total: ${this.rooms.size})`);
        return Message.response({
            action: 'joined_room',
            room_id: roomId,
            room_count: this.rooms.size
        });
    }

    private handleLeaveRoom(payload: Record<string, any>): Message {
        const roomId = payload.room_id || '';
        this.rooms.delete(roomId);
        console.log(`[${this.name}] Left room '${roomId}' (remaining: ${this.rooms.size})`);
        return Message.response({
            action: 'left_room',
            room_id: roomId,
            room_count: this.rooms.size
        });
    }
}

// ============================================
// Chat Application
// ============================================

class ChatApp {
    private roomActor: RoomActor | null = null;
    private sessionActor: SessionActor | null = null;
    private running: boolean = false;

    async start(): Promise<void> {
        // Create actors
        this.roomActor = new RoomActor('general', 'General Chat');
        this.sessionActor = new SessionActor('demo-user', 'Demo User');

        // Start actors
        await this.roomActor.start();
        await this.sessionActor.start();

        // Demo: Auto-join the room
        await this.sessionActor.handleMessage('system', Message.request({
            action: 'join_room',
            room_id: 'general'
        }));

        await this.roomActor.handleMessage('system', Message.request({
            action: 'join',
            user_id: 'demo-user',
            user_name: 'Demo User'
        }));

        this.running = true;
        console.log('=== Aether Chat Application ===');
        console.log("Room: 'General Chat', User: 'Demo User'");
        console.log('Commands: join, leave, send, history, users, info, status, quit');
    }

    async stop(): Promise<void> {
        this.running = false;
        if (this.roomActor) {
            await this.roomActor.stop();
        }
        if (this.sessionActor) {
            await this.sessionActor.stop();
        }
    }

    async demo(): Promise<void> {
        // Wait for startup
        await this.delay(500);

        // Send a welcome message
        if (this.roomActor) {
            await this.roomActor.handleMessage('demo-user', Message.request({
                action: 'send',
                user_id: 'demo-user',
                content: 'Hello, Aether Chat!'
            }));
        }

        await this.delay(300);

        // Get room info
        if (this.roomActor) {
            const info = await this.roomActor.handleMessage('demo-user', Message.request({
                action: 'info'
            }));
            console.log('Room info:', info?.payload);
        }

        await this.delay(300);

        // Get message history
        if (this.roomActor) {
            const history = await this.roomActor.handleMessage('demo-user', Message.request({
                action: 'history',
                limit: 10
            }));
            const payload = history?.payload as Record<string, any> | null;
            console.log('Message count:', payload?.count || 0);
        }

        // Keep running
        while (this.running) {
            await this.delay(1000);
        }
    }

    private delay(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// ============================================
// Main Entry Point
// ============================================

async function main(): Promise<void> {
    const app = new ChatApp();

    // Handle shutdown
    process.on('SIGINT', async () => {
        console.log('\nShutting down...');
        await app.stop();
        process.exit(0);
    });

    try {
        await app.start();
        await app.demo();
    } catch (error) {
        console.error('Application error:', error);
        await app.stop();
        process.exit(1);
    }
}

main();

File: sdks/python/examples/chat_app.py
"""
Chat Application Example

Demonstrates a multi-actor chat room application with room and session actors.
"""
import asyncio
import json
import logging
import time
from typing import Any
from aether_sdk import Actor, Message, MessageType

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class User:
    """Represents a chat user."""
    
    def __init__(self, user_id: str, name: str, joined_at: str = ""):
        self.user_id = user_id
        self.name = name
        self.joined_at = joined_at or time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    
    def to_dict(self) -> dict:
        return {
            "user_id": self.user_id,
            "name": self.name,
            "joined_at": self.joined_at
        }


class ChatMessage:
    """Represents a chat message."""
    
    def __init__(self, msg_id: str, user_id: str, user_name: str, 
                 content: str, room_id: str, timestamp: str = ""):
        self.msg_id = msg_id
        self.user_id = user_id
        self.user_name = user_name
        self.content = content
        self.timestamp = timestamp or time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        self.room_id = room_id
    
    def to_dict(self) -> dict:
        return {
            "id": self.msg_id,
            "user_id": self.user_id,
            "user_name": self.user_name,
            "content": self.content,
            "timestamp": self.timestamp,
            "room_id": self.room_id
        }


class RoomActor(Actor):
    """Manages a single chat room."""
    
    def __init__(self, room_id: str, room_name: str):
        super().__init__(f"room-{room_id}")
        self.room_id = room_id
        self.room_name = room_name
        self.users: dict[str, User] = {}
        self.messages: list[ChatMessage] = []
        self.state_key = f"room_{room_id}_state"
        self.require("STATE_READ", "STATE_WRITE", "ACTOR_MESSAGING", "LOG", "TIME")
    
    async def on_start(self) -> None:
        logger.info(f"[{self.name}] Room '{self.room_name}' starting...")
        
        # Load persisted state
        data = await self.state.read(self.state_key)
        if data:
            try:
                state = json.loads(data)
                for u in state.get("users", []):
                    user = User(u["user_id"], u["name"], u["joined_at"])
                    self.users[user.user_id] = user
                for m in state.get("messages", []):
                    msg = ChatMessage(
                        m["id"], m["user_id"], m["user_name"],
                        m["content"], m["room_id"], m["timestamp"]
                    )
                    self.messages.append(msg)
                logger.info(f"[{self.name}] Restored {len(self.users)} users, {len(self.messages)} messages")
            except Exception as e:
                logger.error(f"[{self.name}] Failed to load state: {e}")
        else:
            await self._save_state()
            logger.info(f"[{self.name}] Initialized new room")
    
    async def on_stop(self) -> None:
        logger.info(f"[{self.name}] Room stopping, saving state...")
        await self._save_state()
    
    async def handle_message(self, sender: str, message: Message) -> Message | None:
        if message.type not in (MessageType.REQUEST, MessageType.RPC_REQUEST):
            return None
        
        payload = message.payload
        if not isinstance(payload, dict):
            return Message.response({"error": "invalid payload"})
        
        action = payload.get("action", "")
        
        if action == "join":
            return await self._handle_join(payload)
        elif action == "leave":
            return await self._handle_leave(payload)
        elif action == "send":
            return await self._handle_send(payload)
        elif action == "history":
            return self._handle_history(payload)
        elif action == "users":
            return self._handle_users()
        elif action == "info":
            return self._handle_info()
        
        return Message.response({"error": f"unknown action: {action}"})
    
    async def _handle_join(self, payload: dict) -> Message:
        user_id = payload.get("user_id", "")
        user_name = payload.get("user_name", "")
        
        if not user_id or not user_name:
            return Message.response({"error": "user_id and user_name required"})
        
        user = User(user_id, user_name)
        self.users[user_id] = user
        await self._save_state()
        
        logger.info(f"[{self.name}] User '{user_name}' joined (total: {len(self.users)})")
        
        return Message.response({
            "action": "joined",
            "room_id": self.room_id,
            "room_name": self.room_name,
            "user_id": user_id,
            "user_count": len(self.users)
        })
    
    async def _handle_leave(self, payload: dict) -> Message:
        user_id = payload.get("user_id", "")
        
        user = self.users.pop(user_id, None)
        if user:
            await self._save_state()
            logger.info(f"[{self.name}] User '{user.name}' left (remaining: {len(self.users)})")
        
        return Message.response({
            "action": "left",
            "room_id": self.room_id,
            "user_id": user_id,
            "user_count": len(self.users)
        })
    
    async def _handle_send(self, payload: dict) -> Message:
        user_id = payload.get("user_id", "")
        content = payload.get("content", "")
        
        if not user_id or not content:
            return Message.response({"error": "user_id and content required"})
        
        user = self.users.get(user_id)
        if not user:
            return Message.response({"error": "user not in room, join first"})
        
        msg = ChatMessage(
            msg_id=f"msg-{int(time.time() * 1000000)}",
            user_id=user_id,
            user_name=user.name,
            content=content,
            room_id=self.room_id
        )
        
        self.messages.append(msg)
        await self._save_state()
        
        logger.info(f"[{self.name}] [{msg.msg_id[:12]}] {user.name}: {content[:30]}...")
        
        return Message.response({
            "action": "sent",
            "message_id": msg.msg_id,
            "message_count": len(self.messages)
        })
    
    def _handle_history(self, payload: dict) -> Message:
        limit = payload.get("limit", 50)
        start = max(0, len(self.messages) - limit)
        messages = [m.to_dict() for m in self.messages[start:]]
        
        return Message.response({
            "action": "history",
            "room_id": self.room_id,
            "messages": messages,
            "count": len(messages)
        })
    
    def _handle_users(self) -> Message:
        users = [u.to_dict() for u in self.users.values()]
        return Message.response({
            "action": "users",
            "room_id": self.room_id,
            "users": users,
            "count": len(users)
        })
    
    def _handle_info(self) -> Message:
        return Message.response({
            "action": "info",
            "room_id": self.room_id,
            "room_name": self.room_name,
            "user_count": len(self.users),
            "message_count": len(self.messages)
        })
    
    async def _save_state(self) -> None:
        state = {
            "room_id": self.room_id,
            "room_name": self.room_name,
            "users": [u.to_dict() for u in self.users.values()],
            "messages": [m.to_dict() for m in self.messages]
        }
        await self.state.write(self.state_key, json.dumps(state).encode())


class SessionActor(Actor):
    """Manages a user session across rooms."""
    
    def __init__(self, user_id: str, user_name: str):
        super().__init__(f"session-{user_id}")
        self.user_id = user_id
        self.user_name = user_name
        self.rooms: set[str] = set()
        self.require("ACTOR_MESSAGING", "LOG")
    
    async def on_start(self) -> None:
        logger.info(f"[{self.name}] Session started for user '{self.user_name}'")
    
    async def on_stop(self) -> None:
        logger.info(f"[{self.name}] Session ended for user '{self.user_name}'")
    
    async def handle_message(self, sender: str, message: Message) -> Message | None:
        if message.type not in (MessageType.REQUEST, MessageType.RPC_REQUEST):
            return None
        
        payload = message.payload
        if not isinstance(payload, dict):
            return Message.response({"error": "invalid payload"})
        
        action = payload.get("action", "")
        
        if action == "status":
            return Message.response({
                "action": "status",
                "user_id": self.user_id,
                "user_name": self.user_name,
                "rooms": list(self.rooms)
            })
        elif action == "join_room":
            room_id = payload.get("room_id", "")
            self.rooms.add(room_id)
            logger.info(f"[{self.name}] Joined room '{room_id}' (total: {len(self.rooms)})")
            return Message.response({
                "action": "joined_room",
                "room_id": room_id,
                "room_count": len(self.rooms)
            })
        elif action == "leave_room":
            room_id = payload.get("room_id", "")
            self.rooms.discard(room_id)
            logger.info(f"[{self.name}] Left room '{room_id}' (remaining: {len(self.rooms)})")
            return Message.response({
                "action": "left_room",
                "room_id": room_id,
                "room_count": len(self.rooms)
            })
        
        return Message.response({"error": f"unknown action: {action}"})


class ChatApp:
    """Main chat application coordinating actors."""
    
    def __init__(self):
        self.room_actor: RoomActor | None = None
        self.session_actor: SessionActor | None = None
        self._running = False
    
    async def start(self) -> None:
        """Start the chat application."""
        # Create actors
        self.room_actor = RoomActor("general", "General Chat")
        self.session_actor = SessionActor("demo-user", "Demo User")
        
        # Start actors
        await self.room_actor.start()
        await self.session_actor.start()
        
        # Demo: Auto-join the room
        await self.session_actor.handle_message("system", Message.request({
            "action": "join_room",
            "room_id": "general"
        }))
        
        await self.room_actor.handle_message("system", Message.request({
            "action": "join",
            "user_id": "demo-user",
            "user_name": "Demo User"
        }))
        
        self._running = True
        logger.info("=== Aether Chat Application ===")
        logger.info("Room: 'General Chat', User: 'Demo User'")
        logger.info("Commands: join, leave, send, history, users, info, status, quit")
    
    async def stop(self) -> None:
        """Stop the chat application."""
        self._running = False
        if self.room_actor:
            await self.room_actor.stop()
        if self.session_actor:
            await self.session_actor.stop()
    
    async def demo(self) -> None:
        """Run a demo of the chat application."""
        await asyncio.sleep(0.5)
        
        # Send a welcome message
        if self.room_actor:
            await self.room_actor.handle_message("demo-user", Message.request({
                "action": "send",
                "user_id": "demo-user",
                "content": "Hello, Aether Chat!"
            }))
        
        await asyncio.sleep(0.3)
        
        # Get room info
        if self.room_actor:
            info = await self.room_actor.handle_message("demo-user", Message.request({
                "action": "info"
            }))
            logger.info(f"Room info: {info.payload}")
        
        await asyncio.sleep(0.3)
        
        # Get message history
        if self.room_actor:
            history = await self.room_actor.handle_message("demo-user", Message.request({
                "action": "history",
                "limit": 10
            }))
            logger.info(f"Message count: {history.payload.get('count', 0)}")
        
        # Keep running
        while self._running:
            await asyncio.sleep(1)


async def main():
    """Run the chat application."""
    app = ChatApp()
    
    try:
        await app.start()
        await app.demo()
    except asyncio.CancelledError:
        logger.info("Application cancelled")
    finally:
        await app.stop()
        logger.info("Application stopped")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.info("Shutting down...")
